libbpf-tools/bitesize.bpf.c

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "bitesize.h"
+
+const volatile char targ_comm[TASK_COMM_LEN] = {};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 10240);
+	__type(key, struct hist_key);
+	__type(value, struct hist);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} hists SEC(".maps");
+
+static struct hist initial_hist;
+
+static __always_inline u64 log2(u32 v)
+{
+	u32 shift, r;
+
+	r = (v > 0xFFFF) << 4; v >>= r;
+	shift = (v > 0xFF) << 3; v >>= shift; r |= shift;
+	shift = (v > 0xF) << 2; v >>= shift; r |= shift;
+	shift = (v > 0x3) << 1; v >>= shift; r |= shift;
+	r |= (v >> 1);
+
+	return r;
+}
+
+static __always_inline u64 log2l(u64 v)
+{
+	u32 hi = v >> 32;
+
+	if (hi)
+		return log2(hi) + 32;
+	else
+		return log2(v);
+}
+
+static __always_inline bool comm_filtered(const char *comm)
+{
+	const volatile char *cond = targ_comm;
+
+	while (*cond) {
+		if (*cond++ != *comm++)
+			return false;
+	}
```

Comment:
I'm surprised this is OK with verifier. It should be considered an infinite loop. Also, on a bit older kernels, where loops are not supported, this will fail. Which kernels have you tested?

I think you should rewrite this as a for loop with up to `TASK_COMM_LEN` iterations, and add `#pragma unroll`

Reply:
I only tested it on `5.7.0-rc7+`, I'll test it on some older kernels, I agree bounded loop with for is better.

Reply:
> I'm surprised this is OK with verifier. It should be considered an infinite loop. Also, on a bit older kernels, where loops are not supported, this will fail. Which kernels have you tested?

@anakryiko Update: I tested on v5.2 ~ v5.5, this is OK from v5.5, I found most of libbpf-tools cann't work on v5.2 ~ v5.4,  for example, this tool will get error on v5.4:

```
libbpf: prog 'tp_btf/block_rq_issue': relo #0: matching candidate #0 request against spec [2473] request + 0:7 => 40.0 @ &x[0].__data_len: 1
libbpf: prog 'tp_btf/block_rq_issue': relo #0: patched insn #36 (LDX/ST/STX) off 40 -> 40
libbpf: block_rq_issue is not found in vmlinux BTF
libbpf: failed to load object 'bitesize_bpf'
libbpf: failed to load BPF skeleton 'bitesize_bpf': -2
failed to load BPF object: -2
```

So we minimum support v5.5 for all libbpf-tools? If so, can we use this kind of loop or we should always use `for` loop with `#pragma unroll`

Reply:
Right, tp_btf programs were added after BPF loop support was added, it seems. So yeah, we don't have to use `#pragma unroll`. But we need to make it a bounded loop anyways (with explicit cap on number of iterations).

Reply:
Get it, thanks. I'll fix these soon.